### PR TITLE
Add 'update' event, and .connecting property public

### PR DESCRIPTION
**Pull request by LuKks**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -48,6 +48,9 @@ Construct a new Hyperswarm instance.
 * `firewall`: A sync function of the form `remotePublicKey => (true|false)`. If true, the connection will be rejected. Defaults to allowing all connections.
 * `dht`: A DHT instance. Defaults to a new instance.
 
+#### `swarm.connecting`
```
Comment (LuKks) [inline README.md:51]: Sounds like a boolean but still ok

Comment (mafintosh) [reply README.md:51]: Its like a boolean++. ie 0, swarm is NOT connecting (false). > 0, it is connecting (true) but you can also inspect how many connecting sockets there are.

File: index.js [hunk 2]
```diff
@@ -146,14 +146,17 @@ module.exports = class Hyperswarm extends EventEmitter {
       relayAddresses: peerInfo.relayAddresses,
       keyPair: this.keyPair
     })
+
+    this.connecting++
+    this.emit('update', 'connecting')
```
Comment (mafintosh) [inline index.js:151]: Move emit to bottom of the function, no reentry bugs then

File: index.js [hunk 3]
```diff
@@ -176,9 +179,11 @@ module.exports = class Hyperswarm extends EventEmitter {
   }
 
   _connectDone () {
-    this._connecting--
-    if (this._connecting < this.maxParallel) this._attemptClientConnections()
-    if (this._connecting === 0) this._flushAllMaybe()
+    this.connecting--
+    this.emit('update', 'connecting')
```
Comment (mafintosh) [inline index.js:183]: Same as above

File: index.js [hunk 4]
```diff
@@ -253,11 +258,15 @@ module.exports = class Hyperswarm extends EventEmitter {
...
 
     this.connections.add(conn)
+    this.emit('update', 'connecting')
+
     this._allConnections.add(conn)
     this._serverConnections++
 
     conn.on('close', () => {
       this.connections.delete(conn)
+      this.emit('update', 'connecting')
```
Comment (mafintosh) [inline index.js:268]: Same as above

File: index.js [hunk 5]
```diff
@@ -165,20 +165,25 @@ module.exports = class Hyperswarm extends EventEmitter {
...
       this.connections.add(conn)
+      this._connectDone()
       conn.removeListener('error', noop)
       peerInfo._connected()
       peerInfo.client = true
       this.emit('connection', conn, peerInfo)
       this._flushMaybe(peerInfo)
     })
+
+    this.connecting++
```
Comment (LuKks) [inline index.js:177]: Should keep this line above where it was before? And just leave the emit in the bottom of the func?

**Comment by mafintosh** [comment]
@kasperisager right now it tracks new connection, disconnect AND when the .connecting prop changes. it will also emit later when the nat info changes (pending the dht-rpc pr)

**Comment by kasperisager** [comment]
@mafintosh Makes sense!

**Comment by LuKks** [comment]
Current limitation: when we receive a server connection then socket is already `'open'`, so we couldn't track `.connecting` for server connections (needs HyperDHT changes?), it's still ok because I think we're more interested on client connections

Btw I will move the `retry timer` test to the other PR where is a bit more relevant

**Comment by LuKks** [comment]
@kasperisager rebased now, in case you would like to confirm

**Review by kasperisager — APPROVED** [review]
Looks good 👍

**Pull request merged**